> the flaky failures on the CI builds keep happening though.

> /home/runner/work/agora/agora/submodules/ocean/src/ocean/text/convert/Formatter.d:466: error: undefined reference to '_D6geod247bitblob__T7BitBlobVmi512ZQp8toStringMxFNfMDFNfAxaZvZv'

It seems it's some kind of linker error.

--- a/source/agora/consensus/SlashPolicy.d
+++ b/source/agora/consensus/SlashPolicy.d
@@ -1,0 +1,289 @@
+/*******************************************************************************
+
+    Manages the slashing policy for the misbehaving validators that do not
+    reveal their pre-images timely.
+
+    This class currently has two responsibilities:
+    - It determines when the misbehaving validators will be slashed -- in other
+        words, how many times of missing pre-images it will allow.
+    - It determines what the penalty is for misbehaving validators -- in other
+        words, How many BOAs it will impose on the misbehaving validators
+        as a penalty
+
+    All the validators should reveal their pre-images timely in order for
+    the network to maintain randomness. So we need a penalty as an incentive
+    to make validators reveal them regularly. So this class manages all the
+    policies for penalty and slashing.
+    See https://github.com/bpfkorea/agora/issues/1076.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.SlashPolicy;
+
+import agora.common.Amount;
+import agora.common.crypto.Key;
+import agora.common.Types;
+import agora.consensus.EnrollmentManager;
+import agora.consensus.data.Params;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.Params;
+import agora.consensus.data.PreImageInfo;
+import agora.consensus.data.Transaction;
+import agora.consensus.state.UTXODB;
+import agora.utils.Log;
+version (unittest) import agora.utils.Test;
+
+import std.exception;
+
+mixin AddLogger!();
+
+/*******************************************************************************
+
+    Manage the policy for slashing
+
+*******************************************************************************/
+
+public class SlashPolicy
+{
+    // The amount of a penalty
+    public immutable Amount penalty_amount;
+
+    // The address to get a penalty
+    public immutable PublicKey penalty_address;
+
+    // Enrollment manager
+    private EnrollmentManager enroll_man;
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            enroll_man = the EnrollmentManager
+            params = the consensus-critical constants
+
+    ***************************************************************************/
+
+    public this (EnrollmentManager enroll_man, immutable(ConsensusParams) params)
+    {
+        this.enroll_man = enroll_man;
+        // TODO: This should be set through ConsensusParams
+        this.penalty_amount = Amount(10_000L * 10_000_000L);
+        this.penalty_address = params.CommonsBudgetAddress;
+    }
+
+    /***************************************************************************
+
+        Get the validators that do not reveal their pre-images and will be
+        slashed through the consensus procoess.
+
+        Params:
+            missing_validators = will contain the validators
+            height = the desired block height to look up the validators for
+
+    ***************************************************************************/
+
+    public void getMissingValidators (ref uint[] missing_validators,
+        in Height height) @safe nothrow
+    {
+        missing_validators.length = 0;
+        () @trusted { assumeSafeAppend(missing_validators); }();
+
+        Hash[] keys;
+        if (!this.enroll_man.getEnrolledUTXOs(keys) || keys.length == 0)
+        {
+            log.fatal("Could not retrieve enrollments / no enrollments found");
+            assert(0);
+        }
+
+        foreach (idx, utxo_key; keys)
+        {
+            if (!this.hasRevealedPreimage(height, utxo_key))
+            {
+                missing_validators ~= cast(uint)idx;
+            }
+        }
+    }
+
+    /***************************************************************************
+
+        Get the random seed reduced from the preimages for the provided
+        block height.
+
+        Params:
+            height = the desired block height to look up the hash for
+
+        Returns:
+            the random seed
+
+    ***************************************************************************/
+
+    public Hash getRandomSeed (in Height height) @safe nothrow
+    {
+        Hash[] keys;
+        if (!this.enroll_man.getEnrolledUTXOs(keys) || keys.length == 0)
+        {
+            log.fatal("Could not retrieve enrollments / no enrollments found");
+            assert(0);
+        }
+
+        Hash[] valid_keys;
+        foreach (key; keys)
+        {
+            if (this.hasRevealedPreimage(height, key))
+                valid_keys ~= key;
+        }
+
+        return this.enroll_man.getRandomSeed(valid_keys, height);
+    }
+
+    /***************************************************************************
+
+        Check if a validator has a pre-image for the height
+
+        Params:
+            height = the desired block height to look up the hash for
+            utxo_key = the UTXO key idendifying a validator
+
+        Returns:
+            true if the validator has revealed its preimage for the provided
+                block height
+
+    ***************************************************************************/
+
+    private bool hasRevealedPreimage (in Height height, in Hash utxo_key)
+        @safe nothrow
+    {
+        auto preimage = this.enroll_man.getValidatorPreimage(utxo_key);
+        auto enrolled = this.enroll_man.getEnrolledHeight(preimage.enroll_key);
+        assert(height >= enrolled);
+        if (preimage.distance >= cast(ushort)(height - enrolled))
+            return true;
+        else
+            return false;
+    }
+}
+
+// Test for getting the candidates to be slashed due to missing pre-images
+unittest
+{
+    import agora.common.crypto.Schnorr;
+    import agora.common.Hash;
+    import agora.consensus.data.Transaction;
+    import agora.consensus.PreImage;
+    import agora.consensus.state.UTXOSet;
+
+    import std.algorithm;
+    import std.format;
+    import std.range;
+
+    scope utxo_set = new TestUTXOSet;
+    Hash[] utxo_hashes;
+
+    // genesisSpendable returns 8 outputs
+    auto pairs = iota(8).map!(idx => WK.Keys[idx]).array;
+    genesisSpendable()
+        .enumerate
+        .map!(tup => tup.value
+            .refund(pairs[tup.index].address)
+            .sign(TxType.Freeze))
+        .each!((tx) {
+            utxo_set.put(tx);
+            utxo_hashes ~= UTXO.getHash(tx.hashFull(), 0);
+        });
+
+    auto params = new immutable(ConsensusParams);
+    scope enroll_man = new EnrollmentManager(":memory:", WK.Keys.A, params);
+
+    // create 8 enrollments
+    Enrollment[] enrollments;
+    PreImageCache[] caches;
+    foreach (idx, kp; pairs)
+    {
+        auto pair = Pair.fromScalar(secretKeyToCurveScalar(kp.secret));
+        auto cycle = PreImageCycle(
+                0, 0,
+                PreImageCache(PreImageCycle.NumberOfCycles, params.ValidatorCycle),
+                PreImageCache(params.ValidatorCycle, 1));
+        const seed = cycle.populate(pair.v, true);
+        caches ~= cycle.preimages;
+        auto enroll = EnrollmentManager.makeEnrollment(
+            pair, utxo_hashes[idx], params.ValidatorCycle,
+            seed, idx);
+        assert(enroll_man.addEnrollment(enroll, kp.address, Height(1),
+                &utxo_set.peekUTXO));
+        enrollments ~= enroll;
+    }
+
+    // 8 validators(=enrollments) are enrolled at the height of 1
+    UTXO[Hash] self_utxos;
+    self_utxos[utxo_hashes[0]] = utxo_set[utxo_hashes[0]];
+    foreach (idx, enroll; enrollments)
+        assert(enroll_man.addValidator(enroll, pairs[idx].address, Height(1),
+            &utxo_set.peekUTXO, self_utxos) is null);
+    assert(enroll_man.validatorCount() == 8);
+
+    // create slashing manager
+    scope slash_man = new SlashPolicy(enroll_man, params);
+
+    // get all the validators and find index of the first and second validastors
+    uint first_validator;
+    uint second_validator;
+    Hash[] utxos;
+    assert(enroll_man.getEnrolledUTXOs(utxos));
+    foreach (idx, utxo; utxos)
+        if (utxo == enrollments[0].utxo_key)
+        {
+            first_validator = cast(uint)idx;
+            break;
+        }
+    foreach (idx, utxo; utxos)
+        if (utxo == enrollments[1].utxo_key)
+        {
+            second_validator = cast(uint)idx;
+            break;
+        }
+
+    // the first validator reveals a pre-image
+    PreImageInfo preimage_1 = PreImageInfo(
+        enrollments[0].utxo_key,
+        caches[0][$ - 2],
+        1
+    );
+    assert(hashFull(preimage_1.hash) == enrollments[0].random_seed);
+    enroll_man.addPreimage(preimage_1);
+    auto gotten_image = enroll_man.getValidatorPreimage(enrollments[0].utxo_key);
+    assert(preimage_1 == gotten_image);
+
+    // the second validator reveals a pre-image
+    PreImageInfo preimage_2 = PreImageInfo(
+        enrollments[1].utxo_key,
+        caches[1][$ - 2],
+        1
+    );
+    assert(hashFull(preimage_2.hash) == enrollments[1].random_seed);
+    enroll_man.addPreimage(preimage_2);
+    gotten_image = enroll_man.getValidatorPreimage(enrollments[1].utxo_key);
+    assert(preimage_2 == gotten_image);
+
+    // check missing pre-image at the current height of 2
+    uint[] missing_validators;
+    slash_man.getMissingValidators(missing_validators, Height(2));
+    assert(missing_validators.length == 6,
+            format!"Current missing preimage validaters: %s"
+                (missing_validators.length));
+    assert(missing_validators.find(first_validator).empty());
+    assert(missing_validators.find(second_validator).empty());
+
+    // get and check random seed
+    Hash preimage_root = slash_man.getRandomSeed(Height(2));
+    assert(preimage_root ==
+        hashMulti(hashMulti(Hash.init, preimage_1), preimage_2));
+}

--- a/source/agora/consensus/SlashPolicy.d
+++ b/source/agora/consensus/SlashPolicy.d
@@ -40,7 +40,8 @@ import agora.consensus.state.UTXODB;
 import agora.utils.Log;
 version (unittest) import agora.utils.Test;
 
-import std.exception;
+import std.algorithm;
+import std.range;
 
 mixin AddLogger!();
 
@@ -121,7 +122,8 @@ public class SlashPolicy
             height = the desired block height to look up the hash for
 
         Returns:
-            the random seed
+            the random seed if there are one or more valid preimages,
+            otherwise Hash.init.
 
     ***************************************************************************/
 
@@ -140,6 +142,11 @@ public class SlashPolicy
             if (this.hasRevealedPreimage(height, key))
                 valid_keys ~= key;
         }
+
+        // NOTE: The preimage root of `Hash.init` value is currently
+        // checked in the `validateSlashingData` function of `Ledger`.
+        if (valid_keys.length == 0)
+            return Hash.init;
 
         return this.enroll_man.getRandomSeed(valid_keys, height);
     }
@@ -168,6 +175,44 @@ public class SlashPolicy
             return true;
         else
             return false;
+    }
+
+    /***************************************************************************
+
+        Check if information for pre-images and slashed validators is valid
+
+        Params:
+            height = the height of proposed block
+            missing_validators = list of indices to the validator UTXO set
+                which have not revealed the preimage
+
+        Returns:
+            `null` if the information is valid at the proposed height,
+            otherwise a string explaining the reason it is invalid.
+
+    ***************************************************************************/
+
+    public string isInvalidPreimageRootReason (Height height,
+        const ref uint[] missing_validators) @safe
+    {
+        Hash[] keys;
+        if (!this.enroll_man.getEnrolledUTXOs(keys) || keys.length == 0)
+        {
+            log.fatal("Could not retrieve enrollments / no enrollments found");
+            assert(0);
+        }
+
+        uint[] local_missing_validators;
+        foreach (idx, key; keys)
+        {
+            if (!this.hasRevealedPreimage(height, key))
+                local_missing_validators ~= cast(uint)idx;
+        }
+
+        if (local_missing_validators != missing_validators)
+            return "The list of missing validators does not match with the local one";
+
+        return null;
     }
 }
 

--- a/source/agora/consensus/protocol/Data.d
+++ b/source/agora/consensus/protocol/Data.d
@@ -24,6 +24,10 @@ public struct ConsensusData
 
     /// The enrollments that are being nominated / voted on
     public Enrollment[] enrolls;
+
+    /// List of indices to the validator UTXO set which have not
+    /// revealed the preimage, aka missing preimage validators, or MPVs
+    public uint[] missing_validators;
 }
 
 /// ConsensusData type testSymmetry check
@@ -55,6 +59,7 @@ unittest
     {
         tx_set:  GenesisBlock.txs,
         enrolls: [ record, record, ],
+        missing_validators : [ 1, 3, 5 ],
     };
 
     testSymmetry(data);

--- a/source/agora/consensus/protocol/Nominator.d
+++ b/source/agora/consensus/protocol/Nominator.d
@@ -280,6 +280,14 @@ extern(D):
             return false;
         }
 
+        // check whether the slashing related data is valid
+        if (auto msg = this.ledger.validateSlashingData(data))
+        {
+            log.fatal("tryNominate(): Invalid preimage data: {}. Data: {}",
+                    msg, data);
+            return false;
+        }
+
         return true;
     }
 
@@ -543,6 +551,14 @@ extern(D):
                 log.error("validateValue(): Validation failed: {}. Data: {}",
                     fail_reason, data);
                 return ValidationLevel.kInvalidValue;
+            }
+
+            if (auto fail_reason = this.ledger.validateSlashingData(data))
+            {
+                log.info("validateValue(): Preimage Validation failed, but " ~
+                    "return kMaybeValidValue. Reason: {}, Data: {}",
+                    fail_reason, data);
+                return ValidationLevel.kMaybeValidValue;
             }
         }
         catch (Exception ex)

--- a/source/agora/node/Ledger.d
+++ b/source/agora/node/Ledger.d
@@ -37,6 +37,7 @@ import agora.consensus.data.Params;
 import agora.consensus.data.Transaction;
 import agora.consensus.state.UTXODB;
 import agora.consensus.EnrollmentManager;
+import agora.consensus.SlashPolicy;
 import agora.consensus.validation;
 import agora.node.BlockStorage;
 import agora.node.Fee;
@@ -77,6 +78,9 @@ public class Ledger
 
     /// Enrollment manager
     private EnrollmentManager enroll_man;
+
+    /// Slashing policy manager
+    private SlashPolicy slash_man;
 
     /// If not null call this delegate
     /// A block was externalized
@@ -120,6 +124,7 @@ public class Ledger
         this.utxo_set = utxo_set;
         this.storage = storage;
         this.enroll_man = enroll_man;
+        this.slash_man = new SlashPolicy(this.enroll_man, params);
         this.pool = pool;
         this.onAcceptedBlock = onAcceptedBlock;
         this.payload_checker = payload_checker;
@@ -439,6 +444,11 @@ public class Ledger
 
         this.enroll_man.getEnrollments(data.enrolls,
             Height(this.getBlockHeight()));
+
+        // get information about validators not revealing a preimage timely
+        this.slash_man.getMissingValidators(data.missing_validators,
+            this.getBlockHeight());
+
         foreach (ref Transaction tx; this.pool)
         {
             if (auto reason = tx.isInvalidReason(utxo_finder, next_height, &this.payload_checker.check))
@@ -492,6 +502,24 @@ public class Ledger
         }
 
         return null;
+    }
+
+    /***************************************************************************
+
+        Check whether the slashing data is valid.
+
+        Params:
+            data = consensus data
+
+        Returns:
+            the error message if validation failed, otherwise null
+
+    ***************************************************************************/
+
+    public string validateSlashingData (in ConsensusData data) @safe
+    {
+        return this.slash_man.isInvalidPreimageRootReason(this.getBlockHeight(),
+                data.missing_validators);
     }
 
     /***************************************************************************
@@ -1377,4 +1405,162 @@ unittest
     blocks = ledger.getBlocksFrom(Height(0)).take(10).array;
     assert(blocks.length == 4);
     assert(blocks[3].header.height == 3);
+}
+
+// create slashing data and check validity for that
+unittest
+{
+    import agora.common.crypto.Schnorr;
+    import agora.consensus.data.genesis.Test;
+    import agora.consensus.data.PreImageInfo;
+    import agora.consensus.PreImage;
+
+    auto params = new immutable(ConsensusParams)(10);
+    const(Block)[] blocks = [ GenesisBlock ];
+    scope ledger = new TestLedger(WK.Keys.NODE2, blocks, params);
+
+    Transaction[] genTransactions (Transaction[] txs)
+    {
+        return txs.enumerate()
+            .map!(en => TxBuilder(en.value).refund(WK.Keys[en.index].address)
+                .sign())
+            .array;
+    }
+
+    Transaction[] genGeneralBlock (Transaction[] txs)
+    {
+        auto new_txs = genTransactions(txs);
+        new_txs.each!(tx => assert(ledger.acceptTransaction(tx)));
+        ledger.forceCreateBlock();
+        return new_txs;
+    }
+
+    // generate payment transaction to the first 8 well-known keypairs
+    auto genesis_txs = genesisSpendable().array;
+    auto txs = genesis_txs[0 .. 4].enumerate()
+        .map!(en => en.value.refund(WK.Keys[en.index].address).sign()).array;
+    txs ~= genesis_txs[4 .. 8].enumerate()
+        .map!(en => en.value.refund(WK.Keys[en.index].address).sign()).array;
+    txs.each!(tx => assert(ledger.acceptTransaction(tx)));
+    ledger.forceCreateBlock();
+    assert(ledger.getBlockHeight() == 1);
+
+    // generate a block with only freezing transactions
+    auto new_txs = txs[0 .. 4].enumerate()
+        .map!(en => TxBuilder(en.value).refund(WK.Keys[en.index].address)
+            .sign(TxType.Freeze)).array;
+    new_txs ~= txs[4 .. 7].enumerate()
+        .map!(en => TxBuilder(en.value).refund(WK.Keys[en.index].address).sign())
+        .array;
+    new_txs ~= TxBuilder(txs[$ - 1]).split(WK.Keys[0].address.repeat(8)).sign();
+    new_txs.each!(tx => assert(ledger.acceptTransaction(tx)));
+    ledger.forceCreateBlock();
+    assert(ledger.getBlockHeight() == 2);
+
+    // UTXOs for enrollments
+    Hash[] utxos = [
+        UTXO.getHash(hashFull(new_txs[0]), 0),
+        UTXO.getHash(hashFull(new_txs[1]), 0),
+        UTXO.getHash(hashFull(new_txs[2]), 0),
+        UTXO.getHash(hashFull(new_txs[3]), 0)
+    ];
+
+    new_txs = iota(new_txs[$ - 1].outputs.length).enumerate
+        .map!(en => TxBuilder(new_txs[$ - 1], cast(uint)en.index)
+            .refund(WK.Keys[en.index].address).sign())
+        .array;
+    new_txs.each!(tx => assert(ledger.acceptTransaction(tx)));
+    ledger.forceCreateBlock();
+    assert(ledger.getBlockHeight() == 3);
+
+    foreach (height; 4 .. 10)
+    {
+        new_txs = genGeneralBlock(new_txs);
+        assert(ledger.getBlockHeight() == Height(height));
+    }
+
+    // add four new enrollments
+    Enrollment[] enrollments;
+    PreImageCache[] caches;
+    auto pairs = iota(4).map!(idx => WK.Keys[idx]).array;
+    foreach (idx, kp; pairs)
+    {
+        auto pair = Pair.fromScalar(secretKeyToCurveScalar(kp.secret));
+        auto cycle = PreImageCycle(
+                0, 0,
+                PreImageCache(PreImageCycle.NumberOfCycles, params.ValidatorCycle),
+                PreImageCache(params.ValidatorCycle, 1));
+        const seed = cycle.populate(pair.v, true);
+        caches ~= cycle.preimages;
+        auto enroll = EnrollmentManager.makeEnrollment(
+            pair, utxos[idx], params.ValidatorCycle,
+            seed, idx);
+        assert(ledger.enroll_man.addEnrollment(enroll, kp.address, Height(1),
+                &ledger.utxo_set.peekUTXO));
+        enrollments ~= enroll;
+    }
+
+    foreach (idx, hash; utxos)
+    {
+        Enrollment stored_enroll = ledger.enroll_man.getEnrollment(hash);
+        assert(stored_enroll == enrollments[idx]);
+    }
+
+    // create the 10th block to make the `Enrollment`s enrolled
+    new_txs = genGeneralBlock(new_txs);
+    assert(ledger.getBlockHeight() == Height(10));
+    ledger.enroll_man.clearExpiredValidators(Height(10));
+    auto b10 = ledger.getBlocksFrom(Height(10))[0];
+    assert(b10.header.enrollments.length == 4);
+
+    // block 11
+    new_txs = genGeneralBlock(new_txs);
+    assert(ledger.getBlockHeight() == Height(11));
+
+    // check missing validators not revealing pre-images.
+    // there are three missing validators at the height of 11.
+    auto temp_txs = genTransactions(new_txs);
+    temp_txs.each!(tx => assert(ledger.acceptTransaction(tx)));
+
+    auto preimage = PreImageInfo(
+        enrollments[0].utxo_key,
+        caches[0][$ - 2],
+        1);
+    ledger.enroll_man.addPreimage(preimage);
+    auto gotten_image =
+        ledger.enroll_man.getValidatorPreimage(enrollments[0].utxo_key);
+    assert(gotten_image == preimage);
+
+    ConsensusData data;
+    ledger.prepareNominatingSet(data, Block.TxsInTestBlock);
+    assert(data.missing_validators.length == 3);
+    assert(data.missing_validators == [0, 1, 2]);
+
+    // check validity of slashing information
+    assert(ledger.validateSlashingData(data) == null);
+    ConsensusData forged_data = data;
+    forged_data.missing_validators = [1, 2, 3];
+    assert(ledger.validateSlashingData(forged_data) != null);
+
+    // reveal preimages of all the validators
+    foreach (idx, cache; caches[1 .. $])
+    {
+        preimage = PreImageInfo(
+            enrollments[idx + 1].utxo_key,
+            cache[$ - 2],
+            1);
+        ledger.enroll_man.addPreimage(preimage);
+        gotten_image =
+            ledger.enroll_man.getValidatorPreimage(enrollments[idx + 1].utxo_key);
+        assert(gotten_image == preimage);
+    }
+
+    // there's no missing validator at the height of 11
+    // after revealing preimages
+    temp_txs.each!(tx => ledger.pool.remove(tx));
+    temp_txs = genTransactions(new_txs);
+    temp_txs.each!(tx => assert(ledger.acceptTransaction(tx)));
+
+    ledger.prepareNominatingSet(data, Block.TxsInTestBlock);
+    assert(data.missing_validators.length == 0);
 }

